go/interpreter/lfvm/instructions.go

```diff
@@ -723,19 +723,51 @@ func opSelfdestruct(c *context) {
 		return
 	}
 
-	gasfunc := gasSelfdestruct
+	beneficiary := tosca.Address(c.stack.pop().Bytes20())
+	// Selfdestruct gas cost defined in EIP-105 (see https://eips.ethereum.org/EIPS/eip-150)
+	cost := tosca.Gas(0)
 	if c.isAtLeast(tosca.R09_Berlin) {
-		gasfunc = gasSelfdestructEIP2929
+		cost += selfDestructedBeneficiaryAccessCost(c.context.AccessAccount(beneficiary))
 	}
+	cost += selfDestructNewAccountCost(c.context.AccountExists(beneficiary),
+		c.context.GetBalance(c.params.Recipient))
 	// even death is not for free
-	if !c.useGas(gasfunc(c)) {
+	if !c.useGas(cost) {
 		return
 	}
-	beneficiary := tosca.Address(c.stack.pop().Bytes20())
-	c.context.SelfDestruct(c.params.Recipient, beneficiary)
+
+	destructed := c.context.SelfDestruct(c.params.Recipient, beneficiary)
+	c.refund += selfDestructRefund(destructed, c.params.Revision)
 	c.status = statusSelfDestructed
 }
 
+func selfDestructedBeneficiaryAccessCost(accessStatus tosca.AccessStatus) tosca.Gas {
+	if accessStatus == tosca.ColdAccess {
+		// Cost of cold account access after EIP 2929 (see https://eips.ethereum.org/EIPS/eip-2929)
+		return 2_600
+	}
+	return 0
+}
+
+func selfDestructNewAccountCost(accountExists bool, balance tosca.Value) tosca.Gas {
+	if !accountExists && balance != (tosca.Value{}) {
+		// cost of creating an account defined in eip-150 (see https://eips.ethereum.org/EIPS/eip-150)
+		// CreateBySelfdestructGas is used when the refunded account is one that does
+		// not exist. This logic is similar to call.
+		// Introduced in Tangerine Whistle (Eip 150)
```

Comment:
The first and last line of this comment say almost the same thing, this could be merged.